### ISSUE TITLE
Drive-by feedback

### ISSUE THREAD

**Issue by domfarolino** [WICG/shared-storage#83]
### Handling inactive/detached documents

The spec right now unconditionally assumes that a Window's browsing context is non-null: [example](https://wicg.github.io/shared-storage/#dom-windowsharedstorage-set), [example](https://wicg.github.io/shared-storage/#dom-windowsharedstorage-append), [example](https://wicg.github.io/shared-storage/#dom-windowsharedstorage-delete), more.

In order to handle the detached frame case:

```js
const win = iframe.contentWindow;
iframe.remove();
win.SharedStorage.set()/append()/foo();
```

... and not crash, we have to bail-out early from algorithms that should not operate in this case. You can do this by detecting the null browsing context as you are, but then you miss the bf-cache case as well, so your best bet is to try and detect non-fully active documents as described by https://w3ctag.github.io/design-principles/#support-non-fully-active. Here is a good example usage: https://w3c.github.io/permissions/#dom-permissions-query of this.

### Nothing defines how the shared storage interface is exposed

At least I don't think anything defines this. The examples in the [Introduction](https://wicg.github.io/shared-storage/#intro) show usage of `window.sharedStorage`, but I don't see the `sharedStorage` member actually being defined on the `Window` interface (if I'm just missing it, forgive me). The closes I see is in https://wicg.github.io/shared-storage/#shared-storage-interface which defines that the interfaces are "exposed" on `Window`, stops short of defining a name and getter. I think you can fix this by adding something similar to https://wicg.github.io/fenced-frame/#window-extension, where we define how the `Fence` interface is exposed via `window.fence`. This would clear up the many lines like this:

> Let context be [WindowSharedStorage](https://wicg.github.io/shared-storage/#windowsharedstorage)'s [Window](https://html.spec.whatwg.org/multipage/nav-history-apis.html#window)'s [browsing context](https://html.spec.whatwg.org/multipage/webappapis.html#window-bc).

... and this:

> Let context be [WorkletSharedStorage](https://wicg.github.io/shared-storage/#workletsharedstorage)'s [SharedStorageWorkletGlobalScope](https://wicg.github.io/shared-storage/#sharedstorageworkletglobalscope)'s [outside settings](https://wicg.github.io/shared-storage/#sharedstorageworkletglobalscope-outside-settings)'s [target browsing context](https://html.spec.whatwg.org/multipage/webappapis.html#concept-environment-target-browsing-context).

... which I don't think are 100% right.

(You might even want to consider doing what we (the fenced frame spec people) should probably do, which is make the getter off of Window just return `null` in the non-fully-active Document case. Then you don't have to keep adding the inactive checks in all of the other individual APIs.)

### Indirect member references

Lines like:

> Let _queue_ be _context_’s associated [shared storage database queue](https://wicg.github.io/shared-storage/#shared-storage-database-shared-storage-database-queue).

Are a tiny bit confusing since there is no member called "shared storage database queue" on browsing context. It might be better to make this:

> Let _queue_ be _context_’s associated [database](https://wicg.github.io/shared-storage/#shared-storage-database)'s [shared storage database queue](https://wicg.github.io/shared-storage/#shared-storage-database-shared-storage-database-queue).

### Bad promise resolution

This includes both _in parallel_ promise resolution, as well as cross-event loop promise resolution. I noticed that in [selectURL()](https://wicg.github.io/shared-storage/#dom-windowsharedstorage-selecturl), we resolve `indexPromise` in an event loop _other than_ the other it was created in, which I don't think is allowed. The closest documentation I found for this is [this example in HTML](https://html.spec.whatwg.org/multipage/webappapis.html#event-loop-for-spec-authors:~:text=Promises%2C%20as%20observable%20JavaScript%20objects%2C%20are%20never%20created%20and%20manipulated%20during%20the%20in%20parallel%20steps.%20p%20is%20created%20before%20entering%20those%20steps%2C%20and%20then%20is%20manipulated%20during%20a%20task%20that%20is%20queued%20specifically%20for%20that%20purpose ) stating that you cannot resolve a promise _in parallel_, but I think more generally the guidance is you must resolve promises in the event loop that they were created in.

Another example of this is in https://wicg.github.io/shared-storage/#dom-workletsharedstorage-append, where while in parallel, step 8 > step 4 resolves the promise, but I don't think this is valid. This must queue a task to the outer event loop per the previous HTML link I referenced. There may be other examples ([delete() seems like one](https://wicg.github.io/shared-storage/#dom-workletsharedstorage-delete)) in your specification where this could be fixed. 

### Misc

 - I noticed in https://wicg.github.io/shared-storage/#report-budget, we define a new member on each "top-level traversable". Just as an FYI, this _excludes_ the top-level browsing context _inside_ of a fenced frame. Is that the intention? In Chrome-implementation-speak, HTML's top-level traversable concept is 1:1 with "primary" frame in the browsing context, whereas "traversable navigable" is any main frame, including ones inside fenced frames. Just wanted to call this out to make sure we thought it through since our terminology is not super duper clear heh :/
 - The https://wicg.github.io/shared-storage/#dom-windowsharedstorage-selecturl algorithm steps seem to be numerically split by the "Issue:" after step 6. Maybe the "Issue:" needs to be indented more?

**Comment by jkarlin** [comment]
@pythagoraskitty PTAL, thanks!

**Comment by pythagoraskitty** [comment]
@domfarolino Thank you so much for the feedback! 

At your convenience, could you PTAL at [PR 85](https://github.com/WICG/shared-storage/pull/85), which attempts to address your feedback?

**Comment by domfarolino** [comment]
Sure, I'm currently on vacation for a bit so I will give it a thorough review mid next week if that is OK. (I don't think the feedback is serious enough to block any intended launch process steps so this spec should be OK from that front).

**Comment by pythagoraskitty** [comment]
Sounds good, thank you!

**Comment by pythagoraskitty** [comment]
This was closed prematurely due to the need to merge the associated PR. Hence, reopening (and will start a new PR to finish  addressing previous comments).

**Comment by pythagoraskitty** [comment]
Linking comments from [previous PR](https://github.com/WICG/shared-storage/pull/85) that I need to address:

- [1](https://github.com/WICG/shared-storage/pull/85#discussion_r1237009074)
- [2](https://github.com/WICG/shared-storage/pull/85#discussion_r1237010268)
- [3](https://github.com/WICG/shared-storage/pull/85#discussion_r1237014575)
- [4](https://github.com/WICG/shared-storage/pull/85#discussion_r1237016883)
- [5](https://github.com/WICG/shared-storage/pull/85#discussion_r1237021550)
- [6](https://github.com/WICG/shared-storage/pull/85#discussion_r1237031502)
- [7](https://github.com/WICG/shared-storage/pull/85#discussion_r1237035130)
- [8](https://github.com/WICG/shared-storage/pull/85#discussion_r1237036891)
- [9](https://github.com/WICG/shared-storage/pull/85#discussion_r1237039741)
- [10](https://github.com/WICG/shared-storage/pull/85#discussion_r1237040665)
